[c10d] Make getDefaultBackend more fault tolerant without relying on exceptions

PatriceVignola opened the pull request:
Summary: no-except builds are terminating when this exception is thrown. We should proactively check if a backend is available before calling has_hooks, instead of trying and failing.

Test Plan: CI

Differential Revision: D71144456




cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroup.hpp:
```diff
@@ -912,15 +912,18 @@ class TORCH_API ProcessGroup : public torch::CustomClassHolder {
           ". Assuming no hooks are registered.");
       return false;
     }
+
+    return getDefaultBackend()->hasHooks();
```
This looks okay, but can we directly use the result of `backendTypeToBackend_.find` to avoid a second search?

kwen2501 reviewed (approved):
Approving. I just have a minor comment for potential of improvement.

PatriceVignola committed:
[c10d] Make getDefaultBackend more fault tolerant without relying on exceptions (#149152)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/149152

no-except builds are terminating when this exception is thrown. We should proactively check if a backend is available before calling has_hooks, instead of trying and failing.

Test Plan: CI

Reviewed By: kwen2501

Differential Revision: D71144456

kwen2501 reviewed (approved):
Nice!

kwen2501 commented:
@pytorchbot merge
